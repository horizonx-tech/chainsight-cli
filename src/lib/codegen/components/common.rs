use std::{fs::OpenOptions, path::Path, io::Read};

use proc_macro2::TokenStream;
use serde::{Deserialize, Serialize};

use crate::types::ComponentType;

#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DatasourceType {
    #[serde(rename = "canister")]
    Canister,
    #[serde(rename = "contract")]
    Contract,
}
#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DestinactionType {
    #[serde(rename = "uint256")]
    Uint256Oracle,
    #[serde(rename = "string")]
    StringOracle,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct Datasource {
    #[serde(rename = "type")]
    pub type_: DatasourceType,
    pub id: String,
    pub method: DatasourceMethod
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct DatasourceMethod {
    pub identifier: String,
    pub interface: Option<String>,
    pub args: Vec<serde_yaml::Value>,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStruct {
    pub name: String,
    pub fields: Vec<DatasourceMethodCustomStructField>
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStructField {
    pub name: String,
    #[serde(rename = "type")]
    pub type_: String
}
#[derive(Clone, Debug, Deserialize, Serialize)]

pub struct DatasourceMethodCustomType {
    pub name: String,
    pub types: Vec<String>
}

impl Datasource {
    // temp: use Default trait
    pub fn default_contract() -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier: "totalSupply():(uint256)".to_string(),
                interface: Some("ERC20.json".to_string()),
                args: vec![],
            },
        }
    }
    pub fn new_contract(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
    
    // temp: use Default trait
    pub fn default_canister(ident_with_ts: bool) -> Self {
        let identifier = if ident_with_ts {
            "get_last_snapshot : () -> (record { value : text; timestamp : nat64 })"
        } else {
            "get_last_snapshot_value : () -> (text)"
        }.to_string();

        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface: None,
                args: vec![],
            },
        }
    }

    pub fn new_canister(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
}

/// Determine indexer type from manifest
pub trait ComponentManifest: std::fmt::Debug {
    fn load(path: &str) -> anyhow::Result<Self> where Self: Sized;
    fn to_str_as_yaml(&self) -> anyhow::Result<String> where Self: Sized;
    fn validate_manifest(&self) -> anyhow::Result<()>;
    fn generate_codes(&self, interface_contract: Option<ethabi::Contract>) -> anyhow::Result<TokenStream>;

    fn component_type(&self) -> ComponentType;
    fn label(&self) -> &str;
    fn destination_type(&self) -> Option<DestinactionType>;
    fn required_interface(&self) -> Option<String>;
}

#[derive(Deserialize)]
pub struct ComponentTypeInManifest {
    #[serde(rename = "type")]
    pub type_: ComponentType,
}
impl ComponentTypeInManifest {
    pub fn determine_type(component_manifest_path: &str) -> anyhow::Result<ComponentType> {
        let mut file = OpenOptions::new()
            .read(true)
            .open(&Path::new(component_manifest_path))?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let data: Self = serde_yaml::from_str(&contents)?;
        Ok(data.type_)
    }
}
