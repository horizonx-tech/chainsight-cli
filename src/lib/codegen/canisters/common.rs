use quote::{format_ident, quote};

use crate::lib::codegen::components::common::{DatasourceMethodCustomStruct, DatasourceMethodCustomType, DatasourceMethodArg};

pub enum OutsideCallIdentsType {
    Eth,
    CrossCanisterCall,
    All
}
/// Generate common identifies such as storage, setter, etc. for outside calls
pub fn generate_outside_call_idents(type_: OutsideCallIdentsType) -> proc_macro2::TokenStream {
    let eth_idents = quote! {
        define_web3_ctx!();
        define_transform_for_web3!();

        manage_single_state!("target_addr", String, false);
    };
    let cross_canister_call_idents = quote! {
        manage_single_state!("target_canister", String, false);
    };
    match type_ {
        OutsideCallIdentsType::Eth => {
            quote! {
                #eth_idents

                setup_func!({
                    target_addr: String,
                    web3_ctx_param: Web3CtxParam
                });
            }
        },
        OutsideCallIdentsType::CrossCanisterCall => {
            quote! {
                #cross_canister_call_idents

                setup_func!({
                    target_canister: String
                });
            }
        },
        OutsideCallIdentsType::All => {
            quote! {
                #eth_idents
                #cross_canister_call_idents

                setup_func!({
                    target_canister: String,
                    target_addr: String,
                    web3_ctx_param: Web3CtxParam
                });
            }
        }
    }
}

// Generate the part of data of the argument that calls the function of datasource contract/canister
pub fn generate_request_arg_idents(method_args: &Vec<DatasourceMethodArg>) -> (Vec<proc_macro2::TokenStream>, Vec<proc_macro2::Ident>) {
    let mut value_idents = vec![];
    let mut type_idents = vec![];
    for method_arg in method_args {
        let DatasourceMethodArg { type_, value } = method_arg;
        // temp
        let request_arg_value = match type_.clone().as_str() {
            "ic_web3::types::U256" => {
                match value {
                    serde_yaml::Value::String(val) => quote! { ic_web3::types::U256::from_dec_str(#val).unwrap() },
                    serde_yaml::Value::Number(val) => {
                        match val.as_u64() {
                            Some(val) => quote! { #val.into() },
                            None => quote! {}
                        }
                    },
                    _ => quote! {}
                }
            }
            "ic_web3::types::Address" => {
                match value {
                    serde_yaml::Value::String(val) => quote! { ic_web3::types::Address::from_str(#val).unwrap() },
                    _ => quote! {}
                }
            },
            _ => {
                match value {
                    serde_yaml::Value::String(val) => {
                        quote! { #val, }
                    },
                    serde_yaml::Value::Number(val) => {
                        match val.as_u64() {
                            Some(val) => {
                                let type_ident = format_ident!("{}", type_);
                                quote! { #val as #type_ident }
                            },
                            None => {
                                quote! {}
                            }
                        }
                    },
                    _ => {
                        quote! {}
                    }
                }
            }
        };
        value_idents.push(request_arg_value);
        if type_ == "ic_web3::types::U256" || type_ == "ic_web3::types::Address" {
            // In the case of contract, other than the primitive type (ic_web3::types::U256 etc.) may be set, in which case type_idents is not used.
            type_idents.push(format_ident!("String")); // temp: thread 'main' panicked at '"ic_web3::types::U256" is not a valid Ident'
        } else {
            type_idents.push(format_ident!("{}", type_));
        }
    };
    (value_idents, type_idents)
}

// Generate CustomStruct Identifiers from manifest's struct data
pub fn generate_custom_struct_idents(custom_structs: &Vec<DatasourceMethodCustomStruct>) -> Vec<proc_macro2::TokenStream> {
    let mut custom_struct_ident: Vec<proc_macro2::TokenStream> = vec![];
    for custom_struct_def in custom_structs {
        let struct_ident = format_ident!("{}", &custom_struct_def.name);
        let mut custom_struct_fields = vec![];
        for field in &custom_struct_def.fields {
            let field_name_ident = format_ident!("{}", &field.name);
            let field_type_ident = format_ident!("{}", &field.type_);
            custom_struct_fields.push(quote! {
                pub #field_name_ident: #field_type_ident,
            });
        }
        custom_struct_ident.push(quote! {
            #[derive(Debug, Clone, candid::CandidType, candid::Deserialize)]
            pub struct #struct_ident {
                #(#custom_struct_fields)*
            }
        });
    };
    custom_struct_ident
}

// Generate CustomType Identifiers from manifest's struct data
pub fn generate_custom_type_idents(custom_types: &Vec<DatasourceMethodCustomType>) -> Vec<proc_macro2::TokenStream> {
    let mut custom_type_ident: Vec<proc_macro2::TokenStream> = vec![];
    for custom_type_def in custom_types {
        let type_ident = format_ident!("{}", &custom_type_def.name);
        let mut custom_type_scalars = vec![];
        for type_ in &custom_type_def.types {
            custom_type_scalars.push(format_ident!("{}", &type_));
        }
        custom_type_ident.push(quote! {
            type #type_ident = (#(#custom_type_scalars),*);
        });
    }
    custom_type_ident
}