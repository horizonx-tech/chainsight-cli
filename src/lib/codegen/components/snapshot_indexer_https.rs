use std::collections::{BTreeMap, HashMap};

use chainsight_cdk::{config::components::CommonConfig, initializer::CycleManagements};
use serde::{Deserialize, Serialize};

use crate::{
    lib::codegen::{canisters, components::common::SourceType, scripts},
    types::{ComponentType, Network},
    utils::serializer::ordered_map,
};

use super::common::{
    custom_tags_interval_sec, ComponentManifest, ComponentMetadata, DestinationType,
    GeneratedCodes, Sources, DEFAULT_MONITOR_DURATION_SECS,
};
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]

pub struct SnapshotIndexerHTTPSDataSource {
    pub url: String,
    #[serde(serialize_with = "ordered_map")]
    pub headers: HashMap<String, String>,
    #[serde(serialize_with = "ordered_map")]
    pub queries: HashMap<String, String>,
}
impl Default for SnapshotIndexerHTTPSDataSource {
    fn default() -> Self {
        Self {
            url: "https://api.coingecko.com/api/v3/simple/price".to_string(),
            headers: vec![("Content-Type".to_string(), "application/json".to_string())]
                .into_iter()
                .collect(),
            queries: vec![
                ("ids".to_string(), "dai".to_string()),
                ("vs_currencies".to_string(), "usd".to_string()),
            ]
            .into_iter()
            .collect(),
        }
    }
}

/// Component Manifest: Snapshot
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct SnapshotIndexerHTTPSComponentManifest {
    #[serde(skip_serializing)]
    pub id: Option<String>,
    pub version: String,
    pub metadata: ComponentMetadata,
    pub datasource: SnapshotIndexerHTTPSDataSource,
    pub interval: u32,
    pub cycles: Option<CycleManagements>,
}

impl SnapshotIndexerHTTPSComponentManifest {
    pub fn new(
        id: &str,
        label: &str,
        description: &str,
        version: &str,
        datasource: SnapshotIndexerHTTPSDataSource,
        interval: u32,
    ) -> Self {
        Self {
            id: Some(id.to_owned()),
            version: version.to_owned(),
            metadata: ComponentMetadata {
                label: label.to_owned(),
                type_: ComponentType::SnapshotIndexerHTTPS,
                description: description.to_owned(),
                tags: Some(vec![
                    "coingecko".to_string(),
                    "DAI".to_string(),
                    "USD".to_string(),
                ]),
            },
            datasource,
            interval,
            cycles: None,
        }
    }
}
impl From<SnapshotIndexerHTTPSComponentManifest>
    for chainsight_cdk::config::components::SnapshotIndexerHTTPSConfig
{
    fn from(val: SnapshotIndexerHTTPSComponentManifest) -> Self {
        let SnapshotIndexerHTTPSComponentManifest { id, datasource, .. } = val;
        Self {
            common: CommonConfig {
                canister_name: id.clone().unwrap(),
                monitor_duration: DEFAULT_MONITOR_DURATION_SECS,
            },
            url: datasource.url,
            headers: BTreeMap::from_iter(datasource.headers),
            queries: BTreeMap::from_iter(datasource.queries),
        }
    }
}

impl ComponentManifest for SnapshotIndexerHTTPSComponentManifest {
    fn load_with_id(path: &str, id: &str) -> anyhow::Result<Self> {
        let manifest = Self::load(path)?;
        Ok(Self {
            id: Some(id.to_owned()),
            ..manifest
        })
    }

    fn to_str_as_yaml(&self) -> anyhow::Result<String> {
        let yaml = serde_yaml::to_string(&self)?;
        Ok(self.yaml_str_with_configs(yaml, "snapshot_indexer_https".to_string()))
    }

    fn validate_manifest(&self) -> anyhow::Result<()> {
        Ok(())
    }

    fn generate_codes(
        &self,
        _interface_contract: Option<ethabi::Contract>,
    ) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::snapshot_indexer_https::generate_codes(self)?,
            types: None,
        })
    }

    fn generate_scripts(&self, network: Network) -> anyhow::Result<String> {
        scripts::snapshot_indexer_https::generate_scripts(self, network)
    }

    fn component_type(&self) -> ComponentType {
        ComponentType::SnapshotIndexerHTTPS
    }

    fn id(&self) -> Option<String> {
        self.id.clone()
    }

    fn metadata(&self) -> &ComponentMetadata {
        &self.metadata
    }

    fn destination_type(&self) -> Option<DestinationType> {
        None
    }

    fn required_interface(&self) -> Option<String> {
        None
    }

    fn generate_user_impl_template(&self) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::snapshot_indexer_https::generate_app(self)?,
            types: None,
        })
    }
    fn get_sources(&self) -> Sources {
        Sources {
            source: self.datasource.url.clone(),
            source_type: SourceType::Https,
            attributes: HashMap::new(),
        }
    }
    fn custom_tags(&self) -> HashMap<String, String> {
        let mut res = HashMap::new();
        let (interval_key, interval_val) = custom_tags_interval_sec(self.interval);
        res.insert(interval_key, interval_val);
        res
    }
    fn cycle_managements(&self) -> Option<CycleManagements> {
        self.cycles.clone()
    }
}

#[cfg(test)]
mod tests {

    use insta::assert_display_snapshot;
    use jsonschema::JSONSchema;

    use crate::lib::test_utils::SrcString;

    use super::*;

    #[test]
    fn test_to_manifest_struct() {
        let yaml = r#"
version: v1
metadata:
    label: sample_snapshot_indexer_https
    type: snapshot_indexer_https
    description: Description
    tags:
    - coingecko
    - DAI
    - USD
datasource:
    url: https://api.coingecko.com/api/v3/simple/price
    headers: 
        content-type: application/json
    queries:
        ids: dai
        vs_currencies: usd
storage:
    with_timestamp: true
interval: 3600
        "#;

        let result = serde_yaml::from_str::<SnapshotIndexerHTTPSComponentManifest>(yaml);
        assert!(result.is_ok());
        let component = result.unwrap();
        assert_eq!(
            component,
            SnapshotIndexerHTTPSComponentManifest {
                id: None,
                version: "v1".to_owned(),
                metadata: ComponentMetadata {
                    label: "sample_snapshot_indexer_https".to_owned(),
                    type_: ComponentType::SnapshotIndexerHTTPS,
                    description: "Description".to_string(),
                    tags: Some(vec![
                        "coingecko".to_string(),
                        "DAI".to_string(),
                        "USD".to_string()
                    ])
                },
                datasource: SnapshotIndexerHTTPSDataSource {
                    url: "https://api.coingecko.com/api/v3/simple/price".to_string(),
                    headers: vec![("content-type".to_string(), "application/json".to_string())]
                        .into_iter()
                        .collect(),
                    queries: vec![
                        ("ids".to_string(), "dai".to_string()),
                        ("vs_currencies".to_string(), "usd".to_string())
                    ]
                    .into_iter()
                    .collect(),
                },
                interval: 3600,
                cycles: None,
            }
        );
        let schema = serde_json::from_str(include_str!(
            "../../../../resources/schema/snapshot_indexer_https.json"
        ))
        .expect("Invalid json");
        let instance = serde_yaml::from_str(yaml).expect("Invalid yaml");
        let compiled = JSONSchema::compile(&schema).expect("Invalid schema");
        let result = compiled.validate(&instance);
        assert!(result.is_ok());
    }

    #[test]
    fn test_snapshot_outputs() {
        let manifest = SnapshotIndexerHTTPSComponentManifest {
            id: Some("sample_snapshot_indexer_https".to_owned()),
            version: "v1".to_owned(),
            metadata: ComponentMetadata {
                label: "Sample Snapshot Indexer Https".to_owned(),
                type_: ComponentType::SnapshotIndexerHTTPS,
                description: "Description".to_string(),
                tags: Some(vec![
                    "coingecko".to_string(),
                    "DAI".to_string(),
                    "USD".to_string(),
                ]),
            },
            datasource: SnapshotIndexerHTTPSDataSource {
                url: "https://api.coingecko.com/api/v3/simple/price".to_string(),
                headers: vec![("content-type".to_string(), "application/json".to_string())]
                    .into_iter()
                    .collect(),
                queries: vec![
                    ("ids".to_string(), "dai".to_string()),
                    ("vs_currencies".to_string(), "usd".to_string()),
                ]
                .into_iter()
                .collect(),
            },
            interval: 3600,
            cycles: None,
        };

        let snap_prefix = "snapshot__snapshot_indexer_https";
        let generated_codes = manifest.generate_codes(Option::None).unwrap();
        assert_display_snapshot!(
            format!("{}__canisters_lib", &snap_prefix),
            SrcString::from(generated_codes.lib)
        );
        assert!(generated_codes.types.is_none());

        let generated_user_impl_template = manifest.generate_user_impl_template().unwrap();
        assert_display_snapshot!(
            format!("{}__logics_lib", &snap_prefix),
            generated_user_impl_template.lib
        );
        assert!(generated_user_impl_template.types.is_none());

        assert_display_snapshot!(
            format!("{}__scripts", &snap_prefix),
            &manifest.generate_scripts(Network::Local).unwrap()
        );
    }
}
