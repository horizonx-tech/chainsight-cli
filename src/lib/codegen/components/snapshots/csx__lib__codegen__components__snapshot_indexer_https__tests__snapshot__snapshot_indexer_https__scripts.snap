---
source: src/lib/codegen/components/snapshot_indexer_https.rs
expression: "&manifest.generate_scripts(Network::Local).unwrap()"
---
#!/bin/bash
# init
dfx canister  call sample_snapshot_indexer_https init_in '(variant { "LocalDevelopment" }, record {
                refueling_interval = 86400: nat64;
                vault_intial_supply = 1000000000000: nat;
                indexer = record { 
                    initial_value = 0: nat;
                    refueling_value = 1000000000000: nat;
                    refueling_threshold = 500000000000: nat;
                };
                db = record { 
                    initial_value = 2000000000000: nat;
                    refueling_value = 100000000000: nat;
                    refueling_threshold = 1000000000000: nat;
                };
                proxy = record { 
                    initial_value = 2000000000: nat;
                    refueling_value = 1000000000: nat;
                    refueling_threshold = 1000000000: nat;
                };
        })' --with-cycles 3002000000000 --wallet $(dfx identity get-wallet )
# set_task
dfx canister  call sample_snapshot_indexer_https set_task '(3600, 10)'

